lib/src/rules/avoid_type_to_string.dart

```diff
@@ -0,0 +1,153 @@
+class AvoidTypeToString extends LintRule implements NodeLintRule {
+  AvoidTypeToString() 
+      : super(
+            name: 'avoid_type_to_string',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(NodeLintRegistry registry, [LinterContext context]) {
+    final visitor = _Visitor(this);
+    registry.addClassDeclaration(this, visitor);
+    registry.addMixinDeclaration(this, visitor);
+    registry.addExtensionDeclaration(this, visitor);
+    registry.addMethodInvocation(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+
+  InterfaceType thisType;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitClassDeclaration(ClassDeclaration node) {
+    // Nodes visited in DFS, so this will be set before each visitMethodInvocation.
+    thisType = node.declaredElement.thisType;
+  }
+
+  @override
+  void visitMixinDeclaration(MixinDeclaration node) {
+    // Nodes visited in DFS, so this will be set before each visitMethodInvocation.
+    thisType = node.declaredElement.thisType;
+  }
+
+  @override
+  void visitExtensionDeclaration(ExtensionDeclaration node) {
+    // Nodes visited in DFS, so this will be set before each visitMethodInvocation.
+    thisType = node.declaredElement.extendedType as InterfaceType;
+  }
+
+  @override
+  void visitMethodInvocation(MethodInvocation node) {
+    final methodName = node.methodName.name;
+    final targetType = (node.realTarget?.staticType is InterfaceType)
+        ?  node.realTarget.staticType as InterfaceType : thisType;
+    final library = node.methodName.staticElement?.library;
+    if (_isToStringOnCoreTypeClass(methodName, targetType, library)) {
+      rule.reportLint(node);
+    }
+
+    node.argumentList.arguments.forEach(_validateArgument);
+  }
+
+  void _validateArgument(Expression expression) {
+    String methodName;
+    InterfaceType targetType;
+    LibraryElement library;
+    if (expression is PropertyAccess) {
+      methodName = expression.propertyName.name;
+      targetType = (expression.realTarget?.staticType is InterfaceType) 
+          ? expression.realTarget?.staticType as InterfaceType : thisType;
+      library = expression.propertyName.staticElement?.library;
+    }
+    else if (expression is SimpleIdentifier) {
+      methodName = expression.name;
+      targetType = thisType;
+      library = expression.staticElement?.library;
+    }
+    else {
+      return;
+    }
+    
+    if (_isToStringOnCoreTypeClass(methodName, targetType, library)) {
+      rule.reportLint(expression);
+    }
+  }
+
+  bool _isToStringOnCoreTypeClass(String methodName, InterfaceType targetType, LibraryElement library) =>
+      methodName == 'toString'
+      && _coreTypeClassInAncestors(targetType)
+      && _rootMethodIsCoreToString(targetType, library);
+
+  bool _coreTypeClassInAncestors(InterfaceType targetType) {
```

Comment:
Consider using the linter context to access the type `Type` (`context.typeProvider.typeType`) and then using the type system (`context.typeSystem`) to ask whether the `targetType` is a subtype of `Type`. That might be slightly cleaner.

Reply:
This was very good advice, thank you.